src/compiler/transformers/classFields.ts

```diff
@@ -664,23 +678,57 @@ namespace ts {
+            if (constructor && constructor.body) {
+                if (useDefineForClassFields) {
+                    statements = statements.filter(statement => !isParameterPropertyDeclaration(getOriginalNode(statement), constructor));
                 }
-                if (afterParameterProperties > indexOfFirstStatement) {
-                    if (!useDefineForClassFields) {
-                        addRange(statements, visitNodes(constructor.body.statements, visitor, isStatement, indexOfFirstStatement, afterParameterProperties - indexOfFirstStatement));
+                else {
+                    for (const statement of constructor.body.statements) {
+                        if (isParameterPropertyDeclaration(getOriginalNode(statement), constructor)) {
+                            parameterPropertyDeclarationCount++;
+                        }
+                    }
+                    if (parameterPropertyDeclarationCount > 0) {
+                        const parameterProperties = visitNodes(constructor.body.statements, visitor, isStatement, indexOfFirstStatementAfterSuper, parameterPropertyDeclarationCount);
+
+                        // If there was a super() call found, add parameter properties immediately after it
+                        if (foundSuperStatement) {
+                            addRange(statements, parameterProperties);
+                        }
+                        // If a synthetic super() call was added, add them just after it
+                        else if (needsSyntheticConstructor) {
+                            statements = addRange(
+                                [statements[0]],
+                                [
+                                    ...parameterProperties,
+                                    ...statements.slice(1),
+                                ]
+                            );
+                        }
+                        // Since there wasn't a super() call, add them to the top of the constructor
+                        else {
+                            statements = addRange(
+                                [],
+                                [
+                                    ...parameterProperties,
+                                    ...statements,
+                                ]
+                            );
+                        }
+
+                        indexOfFirstStatementAfterSuper += parameterPropertyDeclarationCount;
                     }
-                    indexOfFirstStatement = afterParameterProperties;
                 }
             }
-            addPropertyStatements(statements, properties, factory.createThis());
 
-            // Add existing statements, skipping the initial super call.
+            addPropertyStatements(statements, properties, factory.createThis(), getPropertyStatementInsertionIndex(needsSyntheticConstructor, foundSuperStatement, useDefineForClassFields ? 0 : parameterPropertyDeclarationCount));

```

Comment:
Does `getPropertyStatementInsertionIndex` need to be its own function? It seems a conditional would have served just as well:

```suggestion
            const insertionIndex = needsSyntheticConstructor ? 1 : 
                foundSuperStatement ? undefined :
                useDefineForClassFields ? 0 :
                parameterPropertyDeclarationCount;
            addPropertyStatements(statements, properties, factory.createThis(), insertionIndex));
```

Then again, it'd be better to find a way to do this that doesn't use `splice`